### PR TITLE
Ensure column names get implicitly renamed with conflicts

### PR DESCRIPTION
**Pull request by jsternberg**

### CODE REVIEW

**Comment by jsternberg** [comment]
A possible alternative to #5716.

**Comment by jsternberg** [comment]
This is ready to be reviewed.

**Comment by jwilder** [comment]
:+1:

File: influxql/ast.go [hunk 1]
```diff
@@ -1036,34 +1034,70 @@ func (s *SelectStatement) RewriteTimeFields() {
...
 	for _, field := range s.Fields {
+		columnFields = append(columnFields, field)
+
 		switch f := field.Expr.(type) {
 		case *Call:
 			if f.Name == "top" || f.Name == "bottom" {
-				if len(f.Args) == 2 {
-					columnNames = append(columnNames, f.Name)
-					continue
+				for _, arg := range f.Args[1:] {
```
Comment (e-dard) [inline influxql/ast.go]: I assume the capacity of `f.Args` can never be zero?

File: influxql/ast_test.go [hunk 2]
```diff
@@ -1158,6 +1158,57 @@ func Test_fieldsNames(t *testing.T) {
...
+					{Expr: &influxql.VarRef{Val: "total"}, Alias: "value"},
+					{Expr: &influxql.VarRef{Val: "value"}},
+				}),
+			},
+			columns: []string{"time", "value_1", "value", "value_2"},
+		},
+	} {
+		columns := tt.stmt.ColumnNames()
+		if !reflect.DeepEqual(columns, tt.columns) {
+			t.Errorf("%d. expected %s, got %s", i, tt.columns, columns)
```
Comment (e-dard) [inline influxql/ast_test.go]: nit: the error message should be what you got and then what you expected.

**Comment by e-dard** [comment]
one nit and one question, but LGTM :+1:

File: influxql/ast.go (hunk 1 above)
Comment (jsternberg) [inline influxql/ast.go]: I believe the number of args gets verified as part of the `validateAggregates` function so this shouldn't ever be zero.

File: influxql/ast_test.go (hunk 2 above)
Comment (jsternberg) [inline influxql/ast_test.go]: We should probably set some kind of standard for the repository. This package uses expected first everywhere. I see some places in the `tsm1` test files that have got first, but it seems to be more rare.

Comment (e-dard) [inline influxql/ast_test.go]: Yeah I agree. I'm going off of the Go team's [guidelines](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures), which I tend to follow as much as possible.

File: influxql/ast.go (hunk 1 above)
Comment (e-dard) [inline influxql/ast.go]: :+1:

File: influxql/ast_test.go (hunk 2 above)
Comment (jsternberg) [inline influxql/ast_test.go]: I'm ok with switching to that style, but I would like to do it as part of a separate commit for that purpose. I don't want to mix coding style within a single test file if it is avoidable.

Comment (e-dard) [inline influxql/ast_test.go]: OK, suits me. Personally, I've be converting every test file I touch, but
it can mean some significant changes in some files!

On 22 February 2016 at 13:43, Jonathan A. Sternberg <
notifications@github.com> wrote:

> In influxql/ast_test.go
> <https://github.com/influxdata/influxdb/pull/5742#discussion_r53625443>:
>
> > +			columns: []string{"time", "value", "value_1", "value_2"},
> > +		},
> > +		{
> > +			stmt: &influxql.SelectStatement{
> > +				Fields: influxql.Fields([]*influxql.Field{
> > +					{Expr: &influxql.VarRef{Val: "value"}},
> > +					{Expr: &influxql.VarRef{Val: "total"}, Alias: "value"},
> > +					{Expr: &influxql.VarRef{Val: "value"}},
> > +				}),
> > +			},
> > +			columns: []string{"time", "value_1", "value", "value_2"},
> > +		},
> > +	} {
> > +		columns := tt.stmt.ColumnNames()
> > +		if !reflect.DeepEqual(columns, tt.columns) {
> > +			t.Errorf("%d. expected %s, got %s", i, tt.columns, columns)
>
> I'm ok with switching to that style, but I would like to do it as part of
> a separate commit for that purpose. I don't want to mix coding style within
> a single test file if it is avoidable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdata/influxdb/pull/5742/files#r53625443>.
>



-- 
*Edd Robinson*
InfluxDB Engineer
▼▴

*InfluxData.com <http://influxdata.com/>*

Github <http://www.github.com/e-dard> / LinkedIn
<https://www.linkedin.com/in/eddrobinson> / Personal <http://eddrobinson.io>
